tfjs-core/src/util_base.ts

```diff
@@ -749,3 +754,38 @@ export function isPromise(object: any): object is Promise<unknown> {
   //  pre-patched Promise.
   return object && object.then && typeof object.then === 'function';
 }
+
+if (!('setTimeoutWPM' in window)) {
+  const messageName = 'setTimeoutWPM';
+  let fns: Function[] = [];
+  let handledMessageCount = 0;
+
+  // If the setTimeout nesting level is greater than 5 and timeout is less than
+  // 4ms, timeout will be clamped to 4ms, which hurts the perf. Interleaving
+  // window.postMessage and setTimeout will trick the browser and avoid the
+  // clamp.
+  const setTimeoutWPM = function(fn: Function, timeout: number) {
+    fns.push(fn);
+    setTimeout(() => {
+      window.postMessage({name: messageName, index: fns.length - 1}, '*');
+    }, timeout);
+  };
+
+  const handleMessage = function(event: MessageEvent) {
+    if (event.source == window && event.data.name == messageName) {
+      event.stopPropagation();
+      const fn = fns[event.data.index];
+      fn();
+      handledMessageCount++;
+      // console.log(`handledMessageCount=${handledMessageCount},
+      // fns.length=${fns.length}`);
+      if (handledMessageCount === fns.length) {
+        fns = [];
+        handledMessageCount = 0;
+      }
+    }
+  };
+
+  window.addEventListener('message', handleMessage, true);
```

Comment:
It might be better to add this event listener when the user first calls `setTimeoutWPM`. That way, we don't pollute the event listener list if `setTimeoutWPM` is never used.

Reply:
Moved to first call to setTimeoutCustom.